### ISSUE TITLE
Add new imports to aliased module import instead of importing via alias

### ISSUE THREAD

**Issue by djc** [rust-lang/rust-analyzer#16129]
In our project at work we have some imports like this:

```rust
use async_graphql::{self as gql, Object, ID};
```

(This enables the use of `gql::Context<'_>`, whereas having the full `async_graphql::Context<'_>` would be a good deal more verbose.)

However, when we do this, RA will add new imports from the async-graphql crate as:

```diff
  use async_graphql::{self as gql, Object, ID};
+ use gql::InputObject;
```

Instead, the preferred way to do this would be to add it to the original import:

```diff 
- use async_graphql::{self as gql, Object, ID};
+ use async_graphql::{self as gql, Object, ID, InputObject};
```

**rust-analyzer version**: 0.3.1766-standalone (457b966b1 2023-12-10)

**rustc version**: rustc 1.74.1 (a28077b28 2023-12-04)

**relevant settings**:

* Rust-analyzer > Imports > Group: Enable
* Rust-analyzer > Imports > Granularity: Group = `crate`

**Comment by Veykril** [comment]
Hmm, so to add context, the logic for finding an import path is basically optimizng for the shorted path (so `gql` is shorter than `async_graphql` so it picks that prefix).

So the question I see here is, is it always preferred to use the original name in this case, or would there sometimes be the intention of the programmer to only go through the alias for imports (think of a `use dependency as foo;` as a single line at the top).

Though I think in this case, given "Rust-analyzer > Imports > Granularity: Group = `crate`", using the actual crate name does make more sense in these situations.

The logic for the mod path lookups should be this function here https://github.com/veykril/rust-analyzer/blob/96f660813c0f4d6d83e62f3e2de979cc5fe7dd96/crates/hir-def/src/find_path.rs#L317-L329

Unit tests are at the end of the file, I'm not sujre which of the two branches in there is the relevant one, but you can figure that out with a unit test rather easily I think.

**Comment by RalfJung** [comment]
I have a similar wish -- at least it sounds similar so I am adding it here; please let me know if I should file a new issue instead.

It'd be nice if RA could try to preserve the existing import style -- when importing an item from a module that already has imports, extend that import rather than add a new one that names that same module in a different way.
Concretely, say I work in a file at module path `crate::interpret::place` where I already have an import like
```rust
use super::{
    AllocInit, AllocRef, AllocRefMut, CheckAlignMsg, CtfeProvenance, Immediate, InterpCx,
    InterpResult, Machine, MemoryKind, Misalignment, OffsetMode, OpTy, Operand, Pointer,
    Projectable, Provenance, Scalar, alloc_range, interp_ok, mir_assign_valid_types,
};
```
Now say I want to import `ImmTy`, which is in `crate::interpret`, i.e. in the same module that all the other things we're already importing are in. If I ask RA to add that import, it'll add a new line saying
```rust
use crate::interpret::ImmTy;
```
What would be really nice is if RA could realize that the existing `super::` import we have already imports from that same module, so it could just add `ImmTy` to the existing big `use` statement rather than add a new one:
```rust
use super::{
    AllocInit, AllocRef, AllocRefMut, CheckAlignMsg, CtfeProvenance, Immediate, InterpCx,
    InterpResult, Machine, MemoryKind, Misalignment, OffsetMode, OpTy, Operand, Pointer,
    Projectable, Provenance, Scalar, alloc_range, interp_ok, mir_assign_valid_types, ImmTy // <-- new import added here
};
```
It's quite confusing to read code which imports things from the same module using different paths, so having RA prefer already existing imports would help keep code cleaner overall I'd say. :)

I tried this with the default values for `rust-analyzer.imports.granularity.*`. Setting the import granularity to `module` makes no difference.

**Comment by Veykril** [comment]
`rust-analyzer.imports.prefix` is the one that controls this I believe, the default is `crate` which "Force import paths to be absolute by always starting them with `crate` or the extern crate name they come from.`. You want to try either `plain` or `self` there which "Insert import paths relative to the current module, using up to one `super` prefix if the parent module contains the requested item.", in case of `self` it also "Prefixes `self` in front of the path if it starts with a module."

**Comment by ChayimFriedman2** [comment]
It's also difficult for us to infer this unless all imports are relative. If just one uses `super` it's not easy to do.

**Comment by djc** [comment]
Yeah, same here. I also have a project where we have local re-exports of same-workspace dependencies, and Rust-Analyzer will often add new `project_member::module::Item` imports when there is already a `crate::member::module` import line.

**Comment by djc** [comment]
> `rust-analyzer.imports.prefix` is the one that controls this I believe, the default is `crate` which "Force import paths to be absolute by always starting them with `crate` or the extern crate name they come from.`. You want to try either `plain`or`self`there which "Insert import paths relative to the current module, using up to one`super`prefix if the parent module contains the requested item.", in case of`self`it also "Prefixes`self` in front of the path if it starts with a module."

I think the ask here is that R-A honors consider existing import style in the module in which a new import is added. I understand that this is more complex than fixed heuristics, but IMO it would be a strictly better default since it is adaptive.

**Comment by RalfJung** [comment]
I will play around with `rust-analyzer.imports.prefix`, thanks.

However, what I described as ideal behavior is context dependent: scan the existing imports, see if there's one that imports an item from the module that we're adding the import from, if so, extend that instead of adding an entirely new` use`. Only when there is no existing import to extend should it add a new one following `rust-analyzer.imports.prefix`.

> It's also difficult for us to infer this unless all imports are relative. If just one uses super it's not easy to do.

I don't know what you mean by this. RA knows the absolute module path each import refers to, right?

**Comment by ChayimFriedman2** [comment]
> I don't know what you mean by this. RA knows the absolute module path each import refers to, right?

The way r-a finds the path is by searching the crate. We can tell the search to prefer imports via a certain way, but we can't know ahead of time if the existing import can be used.

**Comment by RalfJung** [comment]
Oh, so RA immediately jumps to finding the *path* rather than first finding a suitable *item* and then figuring out which of possibly many equivalent paths to use for that item?

In that framework then I'd phrase this as: first search in modules for which there's already existing imports (i.e., if there's a `use a::b::c`, search in `a::b`), and only if nothing can be found fall back to the current search strategy.

**Comment by ChayimFriedman2** [comment]
No, we do know the item, but items don't keep references to their path.
